middleware/common/include/common/transcode.h

```diff
@@ -17,266 +17,296 @@
-            encode( Iter first, Iter last)
-            {
-               auto cast = []( auto&& i){ return reinterpret_cast< const char*>( &(*i));}; 
-               return detail::encode( cast( first), cast( last));
-            }
+      } // base64
 
-            template< typename Iter>
-            std::enable_if_t< traits::is::binary::iterator_v< Iter>, std::ostream&>
-            encode( std::ostream& out, Iter first, Iter last)
+      namespace utf8
+      {
+         //! @param value String encoded in local default codeset
+         //!
+         //! @return UTF-8-encoded string
+         //!
+         //! @throw exception::limit::Memory on resource failures
+         //! @throw exception::system::invalid::Argument for bad input
+         //! @throw exception::Casual on other failures
+         std::string encode( const std::string& value);
+
+         //! @param value The UTF-8 encoded string
+         //!
+         //! @return String encoded in local default codeset
+         //!
+         //! @throw exception::limit::Memory on resource failures
+         //! @throw exception::system::invalid::Argument for bad input
+         //! @throw exception::Casual on other failures
+         std::string decode( const std::string& value);
+
+         //! @param codeset String-encoding
+         //!
+         //! @return Whether the provided codeset exist in the system
+         //!
+         //! @throw exception::limit::Memory on resource failures
+         //! @throw exception::Casual on other failures
+         bool exist( std::string_view codeset);
+
+         //! @param value String encoded in provided codeset
+         //! @param codeset String-encoding
+         //!
+         //! @return UTF-8-encoded string
+         //!
+         //! @throw exception::limit::Memory on resource failures
+         //! @throw exception::system::invalid::Argument for bad input
+         //! @throw exception::Casual on other failures
+         std::string encode( const std::string& value, std::string_view codeset);
+
+         //! @param value The UTF-8 encoded string
+         //! @param codeset Encoding for result
+         //!
+         //! @return String encoded in provided codeset
+         //!
+         //! @throw exception::limit::Memory on resource failures
+         //! @throw exception::system::invalid::Argument for bad input
+         //! @throw exception::Casual on other failures
+         std::string decode( const std::string& value, std::string_view codeset);
+
+      } // utf8
+
+      namespace hex
+      {
+         namespace detail
+         {
+            template< typename Input, typename Out>
+            void encode( Input first, Input last, Out out)
             {
-               auto cast = []( auto&& i){ return reinterpret_cast< const char*>( &(*i));}; 
-               detail::encode( out, cast( first), cast( last));
-               return out;
-            }
+               auto hex = []( auto value)
```

Comment:
I would not say it's  _obvious_ :) But it's not that difficult, take the lower 4 bits and make a character (we need to adjust to make it a valid hex char, 10 = a, and so on), then take the upper 4 bits.  That is, one byte -> two _chars_ with values [0-f].  
We might be able to use some  c++ _format_ stuff instead of hand rolling it... But I'm yet to find a suitable replacement.